### PR TITLE
libffi 3.4.2

### PR DESCRIPTION
**Pull request by miccal**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

**Comment by miccal** [comment]
Oh dear, for `11-arm64`:
```
Error: 283 failed steps!
```

**Comment by SMillerDev** [comment]
Al lot of them will be fixed with a `libgobject` rebuild

**Comment by miccal** [comment]
> Al lot of them will be fixed with a `libgobject` rebuild

So try a revision bump for `gobject-introspection` as a first step?

**Comment by SMillerDev** [comment]
Yeah, seems like a plan to me

**Comment by carlocab** [comment]
You want to bump anything that's a linkage failure:
```
brew linkage --test duplicity
brew linkage --test ecl
brew linkage --test gforth
brew linkage --test glib
brew linkage --test gobject-introspection
brew linkage --test guile
brew linkage --test inko
brew linkage --test micropython
brew linkage --test moarvm
brew linkage --test ocrmypdf
brew linkage --test p11-kit
brew linkage --test php
brew linkage --test php@7.4
brew linkage --test pygobject3
brew linkage --test rakudo
brew linkage --test txr
brew linkage --test yosys
```
Might be a good tie to switch some of these to system `libffi` on Catalina or newer like so:
https://github.com/Homebrew/homebrew-core/blob/080debca515f613e9fa12daebdba8f29488aecfb/Formula/llvm.rb#L36

`micropython`, for example, is a good candidate, since it depends on `python@3.9`, which uses the system `libffi`

**Comment by miccal** [comment]
Thanks @SMillerDev and @carlocab -- I will wait for the other tests to finish first.

**Comment by carlocab** [comment]
Killed this, since the CI queue has gotten too long, and there's not much more value in waiting for the Catalina tests in addition to the ones on Big Sur and Mojave.

**Comment by miccal** [comment]
`11-arm64` failures:
```
Error: 4 failed steps!
brew test --retry --verbose autogen
brew test --retry --verbose monkeysphere
brew test --retry --verbose include-what-you-use
brew test --retry --verbose easyengine
```

**Comment by SMillerDev** [comment]
`autogen` seems to depend on `guile@2`, which is deprecated but links against `libffi`. We'll have to disable that link, or bump `guile@2`.

**Comment by carlocab** [comment]
`include-what-you-use` needs a revision bump. We should also fix it to not use Cellar paths in its symlinks. I'll push that shortly.

`llvm@{7,8,9}` won't build on Catalina, see #79470, #79468, #79467. The Mojave bottle may not be usable on newer macOS versions (because of the hardcoded `MacOSX10.14.sdk`). If so this will also break `creduce` and `oclgrind`. We could probably make all of them `depends_on maximum_macos: :mojave` and fix them later.

**Comment by SMillerDev** [comment]
Do they not build on Big Sur either? because in that case we're likely better off dropping them completely

**Comment by carlocab** [comment]
They still work on Linux, and they should build on Catalina -- we just need the right `cmake` flag. I just haven't found the time to investigate it.

**Comment by carlocab** [comment]
This will need rebasing if #80413 is merged before this (which I suspect will be the case).

**Comment by miccal** [comment]
Not really sure what to do from here ... ?

**Comment by carlocab** [comment]
Not sure there's much to do at the moment.

I think the various `llvm*` build failures can be resolved in the other currently open PRs, so we should just wait for those. There is also a `gjs` build failure on Mojave that is a bit unexpected:
```
In file included from /private/tmp/gjs--mozjs78-20210701-27088-1kv5pbs/firefox-78.10.1/js/src/gdb/gdb-tests.cpp:9:
In file included from /private/tmp/gjs--mozjs78-20210701-27088-1kv5pbs/firefox-78.10.1/js/src/gdb/gdb-tests.h:23:
In file included from /private/tmp/gjs--mozjs78-20210701-27088-1kv5pbs/firefox-78.10.1/js/src/NamespaceImports.h:15:
In file included from /private/tmp/gjs--mozjs78-20210701-27088-1kv5pbs/firefox-78.10.1/build/dist/include/js/CallNonGenericMethod.h:12:
In file included from /private/tmp/gjs--mozjs78-20210701-27088-1kv5pbs/firefox-78.10.1/build/dist/include/js/CallArgs.h:73:
In file included from /private/tmp/gjs--mozjs78-20210701-27088-1kv5pbs/firefox-78.10.1/build/dist/include/js/RootingAPI.h:27:
/private/tmp/gjs--mozjs78-20210701-27088-1kv5pbs/firefox-78.10.1/build/dist/include/js/ProfilingStack.h:15:10: fatal error: 'js/ProfilingCategory.h' file not found
#include "js/ProfilingCategory.h"
         ^~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

**Comment by SMillerDev** [comment]
Rebasing now that the LLVM changes are in.

**Comment by carlocab** [comment]
I actually have a few more changes I was saving for the weekend when Homebrew/core is quieter, but we can run this now to see how it goes.

**Comment by SMillerDev** [comment]
ARM issue:
```
 travis:
  * Binaries built for an incompatible architecture were installed into travis's prefix.
    The offending files are:
      /opt/homebrew/Cellar/travis/1.10.0_3/libexec/gems/travis-1.10.0/assets/notifications/Travis CI.app/Contents/MacOS/Travis CI
```

Most other issues are audit issues, except for a micropython build failure.

**Comment by SMillerDev** [comment]
Let's try this again and see how far we get.

**Comment by iMichka** [comment]
Only one failure on ARM; that's a good start:
```
2021-08-18T18:35:36.9545540Z [34m==>[0m [1mbrew audit travis --online --git --skip-style[0m
2021-08-18T18:35:50.4790580Z [31m==>[0m [1m[31mFAILED[0m[0m
2021-08-18T18:35:50.4791080Z travis:
2021-08-18T18:35:50.4791840Z   * Binaries built for a non-native architecture were installed into travis's prefix.
2021-08-18T18:35:50.4792440Z     The offending files are:
2021-08-18T18:35:50.4793330Z       /opt/homebrew/Cellar/travis/1.10.0_3/libexec/gems/travis-1.10.0/assets/notifications/Travis CI.app/Contents/MacOS/Travis CI	(x86_64)
2021-08-18T18:35:50.4794300Z [31mError:[0m 1 problem in 1 formula detected
```

**Comment by iMichka** [comment]
`llvm` fails on Linux with:
```
2021-08-18T18:26:54.9778429Z cd /tmp/llvm-20210818-90655-1yug6vw/llvm-project-12.0.1.src/llvm/build/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/libcxx_fuzzer_x86_64-bins/cxx/src && /home/linuxbrew/.linuxbrew/bin/python3.9 /tmp/llvm-20210818-90655-1yug6vw/llvm-project-12.0.1.src/libcxx/utils/merge_archives.py -o /tmp/llvm-20210818-90655-1yug6vw/llvm-project-12.0.1.src/llvm/build/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/libcxx_fuzzer_x86_64/lib/libc++.a --ar /tmp/llvm-20210818-90655-1yug6vw/llvm-project-12.0.1.src/llvm/build/bin/llvm-ar /tmp/llvm-20210818-90655-1yug6vw/llvm-project-12.0.1.src/llvm/build/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/libcxx_fuzzer_x86_64/lib/libc++.a /tmp/llvm-20210818-90655-1yug6vw/llvm-project-12.0.1.src/llvm/build/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/libcxx_fuzzer_x86_64/lib/libc++abi.a 
2021-08-18T18:26:54.9785824Z Traceback (most recent call last):
2021-08-18T18:26:54.9788333Z   File "/tmp/llvm-20210818-90655-1yug6vw/llvm-project-12.0.1.src/libcxx/utils/merge_archives.py", line 11, in <module>
2021-08-18T18:26:54.9790467Z     from ctypes.util import find_library
2021-08-18T18:26:54.9792014Z   File "/home/linuxbrew/.linuxbrew/Cellar/python@3.9/3.9.6/lib/python3.9/ctypes/__init__.py", line 8, in <module>
2021-08-18T18:26:54.9793536Z     from _ctypes import Union, Structure, Array
2021-08-18T18:26:54.9794977Z ImportError: libffi.so.7: cannot open shared object file: No such file or directory
2021-08-18T18:26:54.9798794Z cxx/src/CMakeFiles/cxx_static.dir/build.make:720: recipe for target '/tmp/llvm-20210818-90655-1yug6vw/llvm-project-12.0.1.src/llvm/build/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/libcxx_fuzzer_x86_64/lib/libc++.a' failed
```

**Comment by iMichka** [comment]
`python@3.9`, `python@3.8` and `python@3.7` will need a rev bump too given the above. But let's fix the travis failure before pushing any new change here.

**Comment by carlocab** [comment]
The one failure on ARM is misleading -- there could (and probably will) be more; ARM just skipped the dependent tests because a modified formula failed an audit.

**Comment by iMichka** [comment]
Let's wait for Python@3.9 to be merged to proceed here: #84283

**Comment by SMillerDev** [comment]
Rebased, let's see how far we get.

File: Formula/libffi.rb [hunk 1]
```diff
@@ -17,24 +14,14 @@ class Libffi < Formula
...
 
-  on_macos do
-    if Hardware::CPU.arm?
-      # Improved aarch64-apple-darwin support. See https://github.com/libffi/libffi/pull/565
-      patch do
-        url "https://raw.githubusercontent.com/Homebrew/formula-patches/06252df03c68aee70856e5842f85f20b259e5250/libffi/libffi-3.3-arm64.patch"
-        sha256 "9290aba7f3131ca19eb28fa7ded836b80f15cf633ffac95dc52b14d0a668d1fa"
-      end
-    end
-  end
```
Comment (Bo98) [inline Formula/libffi.rb:36]: Note that the trampoline part of this patch has not been merged yet. Not sure if we have things using that or not.

File: Formula/llvm@7.rb [hunk 2]
```diff
@@ -20,19 +20,21 @@ class LlvmAT7 < Formula
...
+  depends_on maximum_macos: [:catalina, :build]
   depends_on xcode: :build
   depends_on arch: :x86_64
-  depends_on "libffi"
+
+  uses_from_macos "libedit"
+  uses_from_macos "libffi", since: :catalina
+  uses_from_macos "libxml2"
+  uses_from_macos "ncurses"
+  uses_from_macos "zlib"
```
Comment (Bo98) [inline Formula/llvm@7.rb:31]: This should be split into a separate pull request. If there's issues building, ping me and I will look into it.

File: Formula/llvm@8.rb [hunk 3]
```diff
@@ -149,13 +149,19 @@ def install
...
       -DLIBOMP_INSTALL_ALIASES=OFF
     ]
 
+    if MacOS.version >= :catalina
+      args << "-DFFI_INCLUDE_DIR=#{MacOS.sdk_path}/usr/include/ffi"
+      args << "-DFFI_LIBRARY_DIR=#{MacOS.sdk_path}/usr/lib"
+    else
+      args << "-DFFI_INCLUDE_DIR=#{Formula["libffi"].opt_include}"
+      args << "-DFFI_LIBRARY_DIR=#{Formula["libffi"].opt_lib}"
+    end
```
Comment (Bo98) [inline Formula/llvm@8.rb:163]: I'm a little confused how this worked before since this is already `uses_from_macos "libffi"`?

**Comment by SMillerDev** [comment]
Seems `guile`, `guile@2` and php still need a version bump.

**Comment by SMillerDev** [comment]
No idea what's wrong with LLVM.
```console
2021-09-23T09:15:07.9907710Z cd /tmp/llvm-20210923-2338-14gsy0f/llvm-project-12.0.1.src/llvm/build/lib/clang/12.0.1/lib/darwin && codesign --sign - /tmp/llvm-20210923-2338-14gsy0f/llvm-project-12.0.1.src/llvm/build/lib/clang/12.0.1/lib/darwin/libclang_rt.tsan_osx_dynamic.dylib
2021-09-23T09:15:07.9908840Z [ 80%] Built target clang_rt.tsan_osx_dynamic
2021-09-23T09:15:07.9909210Z make[3]: *** [all] Error 2
2021-09-23T09:15:07.9909820Z make[2]: *** [runtimes/runtimes-stamps/runtimes-build] Error 2
2021-09-23T09:15:07.9910490Z make[1]: *** [runtimes/CMakeFiles/runtimes.dir/all] Error 2
2021-09-23T09:15:07.9911000Z make[1]: *** Waiting for unfinished jobs....
```

And PHP fails to build because it uses components that GD removed in the latest release and it somehow wasn't caught in that PR.

**Comment by carlocab** [comment]
> Seems `guile`, `guile@2` and php still need a version bump.

Yea, rebasing will sometimes drop some commits because the same change had been pushed to master.

**Comment by SMillerDev** [comment]
@carlocab any ideas why LLVM could be failing? You seem to know more about that than me.

**Comment by carlocab** [comment]
Nope, no clue. I have seen the error before, but I thought I fixed that.

**Comment by SMillerDev** [comment]
I'll rebase later and see if that might have been the problem.

**Comment by SMillerDev** [comment]
r and fdroidserver need a bump on ARM at least and Linux is still failing to build LLVM. But I'm tempted to just ignore it and see if it works after merging. Because this update has been waiting for too long already iyam.

**Comment by carlocab** [comment]
```
==> brew audit pygobject3 --online --git --skip-style
==> FAILED
pygobject3:
  * stable version should not decrease (from 3.42.0 to 3.40.1)
Error: 1 problem in 1 formula detected
```
Rebasing.

**Comment by SMillerDev** [comment]
ARM passed.... so close 😍

**Comment by SMillerDev** [comment]
I fear we might need to bump python, seeing how sslyze is failing. Either way we need to wait for the letsencrypt fix.

**Comment by carlocab** [comment]
For CI purposes I think letsencrypt should be fixed now. A Python bump will make this PR hit our (long) timeout though, I think.

**Comment by SMillerDev** [comment]
Yeah, that's what I'm fearing too. I'm just not sure how to get around that.

**Comment by carlocab** [comment]
We could make the timeout longer and then change it back after this is merged? `libffi` gets bumped pretty rarely anyway. In the longer term we can split up the formula and dependent tests into separate CI runs so we get double our timeout effectively.

**Comment by SMillerDev** [comment]
I think the timeout is enforced by GitHub

**Comment by carlocab** [comment]
Even on self-hosted runners? Rough. I guess we might have to do my "in the longer term" idea, because I believe GitHub only enforces the timeout on individual jobs rather than the entire workflow.

**Review by carlocab — DISMISSED** [review]
This is going to break things... but I think it was either that or never merge this, so.

**Review by carlocab — APPROVED** [review]
Ok, let's do this before we get bitten by merge conflicts.

**Comment by miccal** [comment]
Thank you all for your hard work with this PR :)

**Pull request closed without merging**